File: api_service/src/services/registry/registry_svc.py
from __future__ import annotations

import contextlib
from time import time
from uuid import UUID

from src.models.registry.instance import ServiceInstance, ServiceSnapshot
from src.repo.etcd_client import (
    EtcdAsyncClient,
    build_prefix,
    build_service_key,
    decode_instance,
    encode_instance,
)
from src.utils.selector_policy import (
    filter_by_tags,
    pick_hash_affinity,
    pick_round_robin,
    random_weighted,
)


class RegistryService:
    """服务注册与发现快照服务。"""

    def __init__(self, etcd_client: EtcdAsyncClient | None = None):
        self._services: dict[str, dict[UUID, ServiceInstance]] = {}
        self._revision = 0
        self._rr_counter = 0
        self._etcd_client = etcd_client
        self._lease_stop_events: dict[str, object] = {}

    async def register(self, instance: ServiceInstance, ttl_sec: int) -> None:
        service_bucket = self._services.setdefault(instance.name, {})
        normalized = ServiceInstance(
            id=instance.id,
            service_id=instance.service_id,
            name=instance.name,
            endpoint=instance.endpoint,
            heartbeat_at=instance.heartbeat_at if instance.heartbeat_at > 0 else time(),
            zone=instance.zone,
            version=instance.version,
            weight=max(instance.weight, 1),
            tags=list(instance.tags),
            active_comm_key_id=instance.active_comm_key_id,
            metadata=dict(instance.metadata),
        )
        service_bucket[instance.id] = normalized
        self._revision += 1

        if self._etcd_client is None:
            return

        key = build_service_key(self._etcd_client.namespace, normalized.name, str(normalized.id))
        ttl = ttl_sec if ttl_sec > 0 else 30
        await self._etcd_client.put_with_lease(key, encode_instance(normalized), ttl=ttl)

    async def unregister(self, instance: ServiceInstance) -> None:
        service_bucket = self._services.get(instance.name)
        if not service_bucket:
            return
        service_bucket.pop(instance.id, None)
        if not service_bucket:
            self._services.pop(instance.name, None)
        self._revision += 1

        if self._etcd_client is None:
            return
        key = build_service_key(self._etcd_client.namespace, instance.name, str(instance.id))
        with contextlib.suppress(Exception):
            await self._etcd_client.delete(key)

    async def get_service_instances(
        self, service_name: str
    ) -> list[ServiceInstance]:
        if self._etcd_client is not None:
            prefix = build_prefix(self._etcd_client.namespace, service_name)
            with contextlib.suppress(Exception):
                items = await self._etcd_client.get_prefix(prefix)
                instances: list[ServiceInstance] = []
                for _, value in items:
                    with contextlib.suppress(Exception):
                        inst = decode_instance(value)
                        instances.append(inst)
                        self._services.setdefault(service_name, {})[inst.id] = inst
                if instances:
                    self._revision += 1
                    return instances

        service_bucket = self._services.get(service_name, {})
        return list(service_bucket.values())

    async def get_service_snapshot(
        self, service_name: str
    ) -> ServiceSnapshot | None:
        instances = await self.get_service_instances(service_name)
        return ServiceSnapshot(
            name=service_name,
            instances=instances,
            revision=self._revision,
        )

    async def choose_endpoint(
        self,
        service_name: str,
        affinity_key: str = "",
        require_tags: list[str] | None = None,
    ) -> ServiceInstance | None:
        instances = await self.get_service_instances(service_name)
        if not instances:
            return None

        now = time()
        live_instances = [
            inst for inst in instances if inst.heartbeat_at <= 0 or now - inst.heartbeat_at <= 30
        ]
        candidates = filter_by_tags(live_instances, require_tags or [])
        if not candidates:
            return None

        if affinity_key:
            return pick_hash_affinity(candidates, affinity_key)

        selected = random_weighted(candidates)
        if selected is not None:
            return selected
        selected = pick_round_robin(candidates, self._rr_counter)
        self._rr_counter += 1
        return selected

File: certification_server/src/services/common/registry_svc.go
package common

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	commonif "certification_server/src/iface/common"
	commonmodel "certification_server/src/models/common"
	modelsystem "certification_server/src/models/system"
	"certification_server/src/repo"
	"certification_server/src/utils"

	"github.com/google/uuid"
	clientv3 "go.etcd.io/etcd/client/v3"
)

const defaultRegistryMaxStaleWindow = 30 * time.Second

var _ commonif.IRegistryManager = (*RegistryService)(nil)

// RegistryService 基于 Etcd 实现服务注册与发现数据源。
type RegistryService struct {
	etcd      *repo.EtcdClient
	keyRoot   string
	opTimeout time.Duration

	leaseMu      sync.Mutex
	leaseIDs     map[string]clientv3.LeaseID
	leaseCancels map[string]context.CancelFunc
	counter      atomic.Uint64
	maxStale     time.Duration
}

// NewRegistryService 创建注册服务。
func NewRegistryService(
	etcdClient *repo.EtcdClient, keyRoot string, opTimeout time.Duration) commonif.IRegistryManager {

	if keyRoot == "" {
		keyRoot = "bms/services"
	}
	if opTimeout <= 0 {
		opTimeout = 3 * time.Second
	}
	return &RegistryService{
		etcd:         etcdClient,
		keyRoot:      strings.Trim(keyRoot, "/"),
		opTimeout:    opTimeout,
		leaseIDs:     make(map[string]clientv3.LeaseID),
		leaseCancels: make(map[string]context.CancelFunc),
		maxStale:     defaultRegistryMaxStaleWindow,
	}
}

// Register 注册服务实例。
func (r *RegistryService) Register(instance *commonmodel.ServiceInstance, ttl int64) error {
	if r.etcd == nil {
		return &modelsystem.ErrNilEtcdClient
	}
	if instance.Name == "" || instance.ID == uuid.Nil {
		return &modelsystem.ErrInstanceNameOrIdRequired
	}
	if instance.HeartBeat <= 0 {
		instance.HeartBeat = time.Now().UnixMilli()
	}
	if instance.Weight <= 0 {
		instance.Weight = 1
	}

	r.stopKeepAlive(instance.Name, instance.ID.String())

	payload, err := json.Marshal(instance)
	if err != nil {
		return fmt.Errorf("%w: %v", &modelsystem.ErrRegistryInstanceMarshal, err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), r.opTimeout)
	defer cancel()

	key := r.instanceKey(instance.Name, instance.ID.String())
	if ttl > 0 {
		_, leaseID, putErr := r.etcd.PutWithTTL(ctx, key, string(payload), ttl)
		if putErr != nil {
			return putErr
		}
		r.startKeepAlive(instance.Name, instance.ID.String(), leaseID, instance)
		return nil
	}

	_, err = r.etcd.Put(ctx, key, string(payload))
	return err
}

// UnRegister 注销服务实例。
func (r *RegistryService) UnRegister(instance *commonmodel.ServiceInstance) error {
	if r.etcd == nil {
		return &modelsystem.ErrNilEtcdClient
	}
	if instance.Name == "" || instance.ID == uuid.Nil {
		return &modelsystem.ErrInstanceNameOrIdRequired
	}

	key := r.instanceKey(instance.Name, instance.ID.String())
	r.stopKeepAlive(instance.Name, instance.ID.String())

	ctx, cancel := context.WithTimeout(context.Background(), r.opTimeout)
	defer cancel()
	_, err := r.etcd.Delete(ctx, key)
	return err
}

// GetServiceInstances 获取服务实例列表。
func (r *RegistryService) GetServiceInstances(serviceName string) ([]*commonmodel.ServiceInstance, error) {
	if r.etcd == nil {
		return nil, &modelsystem.ErrNilEtcdClient
	}
	if serviceName == "" {
		return nil, &modelsystem.ErrServiceNameRequired
	}

	ctx, cancel := context.WithTimeout(context.Background(), r.opTimeout)
	defer cancel()

	resp, err := r.etcd.Get(ctx, r.servicePrefix(serviceName), clientv3.WithPrefix())
	if err != nil {
		return nil, err
	}

	instances := make([]*commonmodel.ServiceInstance, 0, len(resp.Kvs))
	for _, kv := range resp.Kvs {
		var inst commonmodel.ServiceInstance
		if unmarshalErr := json.Unmarshal(kv.Value, &inst); unmarshalErr != nil {
			continue
		}
		instances = append(instances, &inst)
	}

	sort.Slice(instances, func(i, j int) bool { return instances[i].ID.String() < instances[j].ID.String() })
	return instances, nil
}

// GetServiceSnapShot 获取指定服务快照。
func (r *RegistryService) GetServiceSnapShot(serviceName string) (*commonmodel.ServiceSnapshot, error) {
	if r.etcd == nil {
		return nil, &modelsystem.ErrNilEtcdClient
	}
	if serviceName == "" {
		return nil, &modelsystem.ErrServiceNameRequired
	}

	ctx, cancel := context.WithTimeout(context.Background(), r.opTimeout)
	defer cancel()

	resp, err := r.etcd.Get(ctx, r.servicePrefix(serviceName), clientv3.WithPrefix())
	if err != nil {
		return nil, err
	}

	instances := make([]*commonmodel.ServiceInstance, 0, len(resp.Kvs))
	for _, kv := range resp.Kvs {
		var inst commonmodel.ServiceInstance
		if unmarshalErr := json.Unmarshal(kv.Value, &inst); unmarshalErr != nil {
			continue
		}
		instCopy := inst
		instances = append(instances, &instCopy)
	}

	sort.Slice(instances, func(i, j int) bool { return instances[i].ID.String() < instances[j].ID.String() })

	return &commonmodel.ServiceSnapshot{
		Name:      serviceName,
		Instances: instances,
		Revision:  resp.Header.Revision,
	}, nil
}

// ChooseEndpoint 选择服务实例。
func (d *RegistryService) ChooseEndpoint(
	serviceName string, affinityKey string, requireTags []string) (*commonmodel.ServiceInstance, error) {

	if serviceName == "" {
		return nil, &modelsystem.ErrServiceNameRequired
	}

	instances, err := d.GetServiceInstances(serviceName)
	if err != nil {
		return nil, err
	}
	if len(instances) == 0 {
		return nil, &modelsystem.ErrNoAvaliableInstances
	}

	nowMS := time.Now().UnixMilli()
	alive := make([]*commonmodel.ServiceInstance, 0, len(instances))
	for i := range instances {
		inst := instances[i]
		if inst == nil {
			continue
		}
		if inst.HeartBeat <= 0 {
			continue
		}
		if d.maxStale > 0 && nowMS-inst.HeartBeat > d.maxStale.Milliseconds() {
			continue
		}
		alive = append(alive, inst)
	}
	if len(alive) == 0 {
		return nil, &modelsystem.ErrNoAvaliableInstances
	}

	filtered := utils.FilterByTags(alive, requireTags)
	if len(filtered) == 0 {
		return nil, &modelsystem.ErrNoMatchingTags
	}

	var selected *commonmodel.ServiceInstance
	if affinityKey != "" {
		selected = utils.PickHashAffinity(filtered, affinityKey)
	} else {
		selected = utils.RandomWeighted(filtered)
		if selected == nil {
			idx := int(d.counter.Add(1) - 1)
			selected = utils.PickRoundRobin(filtered, idx)
		}
	}

	if selected == nil || selected.Endpoint == "" {
		return nil, &modelsystem.ErrInvalidInstance
	}
	return selected, nil
}

func (r *RegistryService) servicePrefix(serviceName string) string {
	return fmt.Sprintf("/%s/%s/", r.keyRoot, serviceName)
}

func (r *RegistryService) instanceKey(serviceName, instanceID string) string {
	return fmt.Sprintf("/%s/%s/%s", r.keyRoot, serviceName, instanceID)
}

func (r *RegistryService) keepAliveKey(serviceName, instanceID string) string {
	return serviceName + "::" + instanceID
}

func (r *RegistryService) startKeepAlive(
	serviceName,
	instanceID string,
	leaseID clientv3.LeaseID,
	instance *commonmodel.ServiceInstance,
) {
	key := r.keepAliveKey(serviceName, instanceID)
	instanceKey := r.instanceKey(serviceName, instanceID)
	instanceTemplate := commonmodel.ServiceInstance{}
	if instance != nil {
		instanceTemplate = *instance
	}

	r.leaseMu.Lock()
	if oldCancel, ok := r.leaseCancels[key]; ok {
		oldCancel()
	}
	ctx, cancel := context.WithCancel(context.Background())
	r.leaseIDs[key] = leaseID
	r.leaseCancels[key] = cancel
	r.leaseMu.Unlock()

	ch, err := r.etcd.KeepAlive(ctx, leaseID)
	if err != nil {
		cancel()
		r.leaseMu.Lock()
		delete(r.leaseIDs, key)
		delete(r.leaseCancels, key)
		r.leaseMu.Unlock()
		return
	}

	go func(k string, c <-chan *clientv3.LeaseKeepAliveResponse) {
		for range c {
			if instanceTemplate.ID != uuid.Nil {
				refresh := instanceTemplate
				refresh.HeartBeat = time.Now().UnixMilli()
				if payload, marshalErr := json.Marshal(&refresh); marshalErr == nil {
					putCtx, putCancel := context.WithTimeout(context.Background(), r.opTimeout)
					_, _ = r.etcd.Put(putCtx, instanceKey, string(payload), clientv3.WithLease(leaseID))
					putCancel()
				}
			}
		}
		r.leaseMu.Lock()
		delete(r.leaseIDs, k)
		delete(r.leaseCancels, k)
		r.leaseMu.Unlock()
	}(key, ch)
}

func (r *RegistryService) stopKeepAlive(serviceName, instanceID string) {
	key := r.keepAliveKey(serviceName, instanceID)

	r.leaseMu.Lock()
	leaseID, hasLease := r.leaseIDs[key]
	cancel, hasCancel := r.leaseCancels[key]
	delete(r.leaseIDs, key)
	delete(r.leaseCancels, key)
	r.leaseMu.Unlock()

	if hasCancel {
		cancel()
	}
	if hasLease {
		ctx, c := context.WithTimeout(context.Background(), r.opTimeout)
		_ = r.etcd.RevokeLease(ctx, leaseID)
		c()
	}
}
